protocol/messenger_handler.go

```diff
@@ -683,6 +684,63 @@ func (m *Messenger) HandleCommunityInvitation(state *ReceivedMessageState, signe
 	return nil
 }
 
+func (m *Messenger) HandleHistoryArchiveMagnetlinkMessage(state *ReceivedMessageState, communityPubKey *ecdsa.PublicKey, magnetlink string, clock uint64) error {
+
+	id := types.HexBytes(crypto.CompressPubkey(communityPubKey))
+	settings, err := m.communitiesManager.GetCommunitySettingsByID(id)
+	if err != nil {
+		m.logger.Debug("Couldn't get community settings for community with id: ", zap.Any("id", id))
+		return err
+	}
+
+	if m.config.torrentConfig != nil && m.config.torrentConfig.Enabled && settings.HistoryArchiveSupportEnabled {
+		signedByOwnedCommunity, err := m.communitiesManager.IsAdminCommunity(communityPubKey)
+		if err != nil {
+			return err
+		}
+		joinedCommunity, err := m.communitiesManager.IsJoinedCommunity(communityPubKey)
+		if err != nil {
+			return err
+		}
+		lastClock, err := m.communitiesManager.GetMagnetlinkMessageClock(id)
+		if err != nil {
+			return err
+		}
+		// We are only interested in a community archive magnet link
+		// if it originates from a community that the current account is
+		// part of and doesn't own the private key at the same time
+		if !signedByOwnedCommunity && joinedCommunity && clock >= lastClock {
+
+			m.communitiesManager.UnseedHistoryArchiveTorrent(id)
+			go func() {
+				downloadedArchiveIDs, err := m.communitiesManager.DownloadHistoryArchivesByMagnetlink(id, magnetlink)
+				if err != nil {
+					log.Println("failed to download history archive data", err)
+					m.logger.Debug("failed to download history archive data", zap.Error(err))
+					return
+				}
+
+				messagesToHandle, err := m.communitiesManager.ExtractMessagesFromHistoryArchives(id, downloadedArchiveIDs)
+				if err != nil {
+					log.Println("failed to extract history archive messages", err)
+					m.logger.Debug("failed to extract history archive messages", zap.Error(err))
+					return
+				}
+
+				_, err = m.handleRetrievedMessages(messagesToHandle, false)
```

Comment:
This looks a bit dangerous, we currently don't run `handleRetrieveMessages` in parallel, so weird things might happen

Reply:
Hm.. I wonder what such weird things could be, given that `handleRetrieveMessages` has various checks for incoming messages and handling out of order messages etc.

Any suggestions on how else this should be done?

Reply:
First thing that comes to mind is to have a channel, and handle retrieved messages it's called on it, so you have 2 producers (polling from the network, and torrent) and a single consumer, a bit more complex, but we don't have to think about the code running in parallel, as it might lead to difficult to find bugs.

Reply:
Okay I'll see if I can make that happen. Thanks for your input!

Reply:
You also need to publish the response, otherwise the client will be in inconsistent state, https://github.com/status-im/status-go/blob/f66ad3d3dd754a8c2e789e3339c5bf0b970129c1/services/ext/service.go#L210 , have you tested this on desktop to see if changes are reflected on the frontend?

Reply:
In desktop, what we do is we listen for the `HistoryDownloaded` signal, and if that has happened we refetch messages from status-go for the current active channel.